src/couch_stats/src/couch_stats_resource_tracker.erl

```diff
@@ -0,0 +1,912 @@
+    %% Only return non zero (and also positive) delta fields
+    maps:filter(fun(_K,V) -> V > 0 end, Delta);
+make_delta(_, #rctx{}) ->
+    #{error => missing_beg_rctx};
+make_delta(#rctx{}, _) ->
+    #{error => missing_fin_rctx}.
+
+%% TODO: what to do when PidRef=undefined?
+make_delta_base(PidRef) ->
+    %% TODO: extract user_ctx and db/shard from request
+    Now = tnow(),
+    #rctx{
+        pid_ref = PidRef,
+        %% TODO: confirm this subtraction works
+        started_at = Now - 100, %% give us 100ms rewind time for missing T0
+        updated_at = Now
+    }.
+
+make_delta_base() ->
+    make_delta_base(get_pid_ref()).
+
+set_delta_a(TA) ->
+    erlang:put(?DELTA_TA, TA).
+
+update_counter(Field, Count) ->
+    is_enabled() andalso update_counter(get_pid_ref(), Field, Count).
+
+update_counter(undefined, _Field, _Count) ->
+    ok;
+update_counter({_Pid,_Ref}=PidRef, Field, Count) ->
+    %% TODO: mem3 crashes without catch, why do we lose the stats table?
+    is_enabled() andalso catch ets:update_counter(?MODULE, PidRef, {Field, Count}, #rctx{pid_ref=PidRef}).
+
+update_element(undefined, _Update) ->
+    ok;
+update_element({_Pid,_Ref}=PidRef, Update) ->
+    %% TODO: should we take any action when the update fails?
+    is_enabled() andalso catch ets:update_element(?MODULE, PidRef, Update).
+
+%% Process lifetime logging api
+
+track(#rctx{pid_ref=PidRef}) ->
+    case get_tracker() of
+        undefined ->
+            Pid = spawn(?MODULE, tracker, [PidRef]),
+            put_tracker(Pid),
+            Pid;
+        Pid when is_pid(Pid) ->
+            Pid
+    end.
+
+tracker({Pid, _Ref}=PidRef) ->
+    MonRef = erlang:monitor(process, Pid),
+    receive
+        stop ->
+            %% TODO: do we need cleanup here?
+            log_process_lifetime_report(PidRef),
+            catch evict(PidRef),
+            demonitor(MonRef),
+            ok;
+        {'DOWN', MonRef, _Type, _0DPid, _Reason0} ->
+            destroy_context(PidRef),
+            %% TODO: should we pass reason to log_process_lifetime_report?
+            %% Reason = case Reason0 of
+            %%     {shutdown, Shutdown0} ->
+            %%         Shutdown = atom_to_binary(Shutdown0),
+            %%         <<"shutdown: ", Shutdown/binary>>;
+            %%     Reason0 ->
+            %%         Reason0
+            %% end,
+            %% TODO: should we send the induced work delta to the coordinator?
+            log_process_lifetime_report(PidRef),
+            catch evict(PidRef)
+    end.
+
+log_process_lifetime_report(PidRef) ->
+    case is_enabled() andalso is_logging_enabled() of
+        true ->
+            Rctx = get_resource(PidRef),
+            case should_log(Rctx) of
+               true ->
+                    couch_log:report("csrt-pid-usage-lifetime", to_json(Rctx));
+                _ ->
+                    ok
+            end;
+        false ->
+            ok
+    end.
+
+is_logging_enabled() ->
```

Comment:
Please add a comment here that `is_logging_enabled()` suppose to be `true` for both `true` and `coordinator`.

Reply:
Alright, I hacked together a `csrt_logger` module in 17021155d that provides two extended mechanisms for dynamically enabling logging, in both a simpler and more complex manner. I just hacked this together, there's no compiler warnings but I haven't actually run it yet or anything wild like that.

The first dynamic logging approach utilizes @iilyak's idea for dedicated type records, initially focusing on `#worker{}` type here https://github.com/apache/couchdb/commit/17021155db6ae0495e015d7571469a36c426b894#diff-ae4228aa19a3865ec48888021914ba2db0cf64dca17138497c8db885222f4699R77-R90 eg `#rpc_worker{mod=fabric_rpc, func=map_view}`. We then translate that to a key string like `worker.fabric_rpc.map_view` so we can then do `config:get_boolean("csrt.should_log", Key, true)` to have a direct toggle switch for the core default rpc operations we want to log, and we could use a similar mechanism to dynamically enable logging for additional rpc endpoints we don't enable by default, eg looking at dreyfus_rpc or mem3_rpc.

We generate that mapping in a gen_server so that we can easily do direct atom comparisons on module/function names rather than have to repeatedly do string conversion and comparisons, the strings are only changed at configuration initialization/update time and not used in the should log checks. We can then check to see if that module worker type is tracked directly https://github.com/apache/couchdb/commit/17021155db6ae0495e015d7571469a36c426b894#diff-ae4228aa19a3865ec48888021914ba2db0cf64dca17138497c8db885222f4699R220-R221

The next mechanism and core idea with this experimental hack in the commit is to create a way to dynamically register match spec funs to evaluate against the stream of reports. The idea here is to `ets:match_spec_compile` to generate a compiled match_spec we can store a reference to in the gen_server, then we can dynamically enable filtered logging by way of declaring and registering match_specs. For example, here's our matcher registration handling https://github.com/apache/couchdb/commit/17021155db6ae0495e015d7571469a36c426b894#diff-ae4228aa19a3865ec48888021914ba2db0cf64dca17138497c8db885222f4699R171-R174 along with the compilation https://github.com/apache/couchdb/commit/17021155db6ae0495e015d7571469a36c426b894#diff-ae4228aa19a3865ec48888021914ba2db0cf64dca17138497c8db885222f4699R209-R218 and then we can determine if we should logic the particular `#rctx{}` by way of checking if it matches any of the registered filteres here https://github.com/apache/couchdb/commit/17021155db6ae0495e015d7571469a36c426b894#diff-ae4228aa19a3865ec48888021914ba2db0cf64dca17138497c8db885222f4699R133-R142

For example, here's a function to dynamically generate a matchspec parameterized by username:

```erlang
(node1@127.0.0.1)4> F = fun(User) -> ets:fun2ms(fun(#rctx{username=User0}=R) when User =:= User0 -> R end) end.
#Fun<erl_eval.42.3316493>
(node1@127.0.0.1)5> F(bar).
[{#rctx{started_at = '_',updated_at = '_',pid_ref = '_',
        mfa = '_',nonce = '_',from = '_',type = '_',dbname = '_',
        username = '$1',path = '_',db_open = '_',docs_read = '_',
        rows_read = '_',changes_processed = '_',
        changes_returned = '_',ioq_calls = '_',io_bytes_read = '_',
        io_bytes_written = '_',js_evals = '_',js_filter = '_',
        js_filtered_docs = '_',mango_eval_match = '_',
        get_kv_node = '_',get_kp_node = '_',write_kv_node = '_',
        write_kp_node = '_'},
  [{'=:=',{const,bar},'$1'}],
  ['$_']}]
```

Then we can compile that match spec and store the reference in the gen_server, and check to see if any of the compiled matchers match the given `#rctx{}` record. This opens up considerable options for doing dynamic filtered logging of a subset of the csrt report stream, although if we adopt a model like this I think this would be great to extend out to the main logging systems as well, so for instance we could dynamically enable debug logs for a particular database shard. I think it might be trickier to dynamically define relations, eg `username=Foo and docs_opened>100`, but it _is_ easy to predefine a series of those types of functions for common use cases like filter on user and doc magnitude, or user and db shard plus magnitude, or whatever. Perhaps there's some other clever tricks we can do here too, it's a little tricky with having to define the match record within the macro expansion call, but as I demonstrated above we can dynamically generate parameterized match specs that are then compiled and shareable, so I think there's a lot of potential with this current approach.

Reply:
> The idea here is to ets:match_spec_compile to generate a compiled match_spec we can store a reference to in the gen_server, then we can dynamically enable filtered logging by way of declaring and registering match_specs. 

A combination of ` ets:match_spec_compile` and `ets:match_spec_run` is a very neat idea. Did you consider https://www.erlang.org/doc/apps/erts/persistent_term.html instead of gen_server? I wish persistent_term had `persistent_term:from_map/1`.

Reply:
I looked briefly into bif implementation. It is nice:

- hash map internally
- the term sharing is preserved for complex containers (tuples, lists)
- the literals like numbers and strings are copied
- only one allocation during `get` (to create reference for GC)

My conclusion is - it should be way faster than `ets` or `gen_server` based approach. When updates are rare. Updates causing locking of a data structure so all concurrent reads are on hold.

Reply:
Ahhhhh that's an interesting idea with using persistent_term, I need to remember that's a thing now.

The purpose of the gen_server was two fold: 1) to have a way of registering and preserving the filters; 2) to have a mechanism for getting the registered filters applied to the reports as the the CSRT contexts are cleaned up.

There's various ways to do 1), but for 2) my experiment above used a gen_server so that it could store the filters and then have the logs funneled through the gen_server, which likely should use sharded gen_servers like we've done elsewhere.

This isn't strictly necessary though, as the current PR approach uses the dedicated monitor process to generate the report https://github.com/apache/couchdb/blob/couch-stats-resource-tracker-v2/src/couch_stats/src/couch_stats_resource_tracker.erl#L734 and avoids needing dedicated servers to filter the operations. The tricky bit was about how to get the registered filters down to the monitor processes so they could determine locally whether to log the report or not.

The use of persistent terms is an interesting approach for declaring the filter matchers, especially since the output `ets:match_spec_compile` is a reference to an internal representation, so fetching the set of references across the various workers and coordinators to then apply locally seems worth exploring more, especially since it allows us to stick with the decentralized monitor processes rather than creating new high throughput gen_server's.

We still have the aspect of discovery, I'm curious what you had in mind on this front @iilyak, but from the persistent term API unless we use some other mechanism for dynamically declaring the set of enabled filters, we'll have to do something with an expected naming scheme and iterate over the entries. Maybe something like

```erlang
register_filter(Name, CompMatchSpec) when is_binary(Name0) ->
    Name = <<"csrt-logging-filter::", Name0/binary>>,
    persistent_term:put(Name, CompMatchSpec).

get_active_filters() ->
    lists:filter(fun (<<"csrt-logging-filter::", _/Binary>>) -> true; (_) -> false end, persistent_term:get()).
```

Basically using https://www.erlang.org/doc/apps/erts/persistent_term.html#get/0 to iterate over all declared persistent terms and then looking for an expected prefix naming scheme. Did you have something else in mind? I suppose we could also have a gen_server for handling the state and declarations, but there still needs to be some mechanism for publishing what is active. That might also be useful if we want to have the ability to register TTL filtering, but I'm not convinced that's worth complicating the design over. I'm curious what thoughts you had on this front, how would a given CSRT monitor process query the registered filters?

Hmmm... I suppose we could make it even simpler and just do `persistent_term:put(all_the_filters, #{...}` and then just store and fetch a map of those entries. That might be a decent idea, we need to iterate through _all_ the filters so we might as well grab them in one query. That also avoids having to do the dynamic naming sequences. Then our earlier functions become:

```erlang
register_filter(Name, CompMatchSpec) ->
    Filters = persistent_term:get(all_the_filters, #{}),
    persistent_term:put(all_the_filters, maps:put(Name, CompMatchSpec, Filters)).

get_active_filters() ->
    persistent_term:get(all_the_filters, #{}).
```